### ISSUE TITLE
MeasureText appears to return wrong values

### ISSUE THREAD

**Issue by peardox** [skia4delphi/skia4delphi#340]
Consider this...

```
var SpaceRect: TRectF;
var LFont1: ISkFont;
LFont1 := TSkFont.Create(TSkTypeface.MakeFromFile('RobotoMono-Regular.ttf'), 32);
LFont1.MeasureText(' ', SpaceRect);

```

i.e. Measurung a Space char. After this call SpaceRect is Empty (all zero) and while it's true that a space is empty it does occupy some area so the TRectF should be > 0

[Skia API](https://api.skia.org/classSkFont.html#ab50421ba467b68bebac4282f6057cfc5) implies something is wrong...

**Comment by viniciusfbb** [comment]
@peardox I think it is working as expected, or at least the issue is on Google Skia side. I made a small example on Google Skia (removing Skia4Delphi from the table just for test purpose): https://fiddle.skia.org/c/a58d7fb3f863768a7b5ee622f2019e73 and the output is `left:0  top:0  right:0  bottom:0`

Text manipulations in general are complex as there are several variants, like WordWrap, Lines, Trimming, MaxWidth, MultipleStyles, RightToLeft, HorizontalAlignment (including justified), Locale and much more. There are a couple of text apis on Google Skia, and the most robust (and recommended) is the SkParagraph.

I really recommend you to use the SkParagraph in case of Vcl or FMX.Skia.Canvas.TSkTextLayout in case of FMX (that use SkParagraph but in a simple way).

**Comment by peardox** [comment]
Thanks

I've got a partial solution using something like ...
  APosBuf := LFont1.GetHorizontalPositions(LGlyphs); // LGlyphs is a sequential list of character set codes e.g. 32..127

The returned array gives the correct value for a space

I can't use any Paragraph or TestLayout stuff as I'm trying to create bitmaps from font glyphs (for microcontrollers to display)

Today I discovered that sometimes the vertical pixels of adjacent chars can overlap e.g. i+j in some fonts the end of j is direclty under the i. Going to examine if GetBounds can help next...

There's nothing I can find in Skia to do with kerning (which may have helped)

Got it mostly working with a few niggling little issues to resolve like above
